### ISSUE TITLE
Stack overflow when parsing an expression

### ISSUE THREAD

**Issue by orf** [Instagram/LibCST#1000]
Below I've attached a real file that has been published on PyPI. It contains an impressively gigantic expression.

[huge.txt](https://github.com/Instagram/LibCST/files/12445264/huge.txt)

Parsing this file with the current libcst master branch causes a stack overflow. After some debugging, this appears to be caused by the `cache_left_rec` macro cloning the expression ([here](https://github.com/kevinmehall/rust-peg/blob/c870876adc9a9601f14dfa3e13c5a691ddfeb8ee/peg-macros/translate.rs#L292-L294)), which in this case causes a _gigantic_ recursive clone of the expression, eventually overflowing the stack.

<img width="840" alt="Screenshot 2023-08-26 at 12 40 04" src="https://github.com/Instagram/LibCST/assets/1027207/175eb80c-7413-4853-9b39-6df7fa2e2e17">

**Comment by zsol** [comment]
FWIW CPython also chokes on this input:
```
❯ py -m py_compile 'huge.txt'
Sorry: RecursionError: maximum recursion depth exceeded during compilation
```

**Comment by orf** [comment]
Yeah, it's definitely a pathalogical file. What's annoying is that this brings down the entire process in Rust and there isn't _really_ a way to stop that from happening, so you sort of need to heuristically detect if a given file will stack overflow or not based on the tokens.

I thought I'd add this issue anyway.

**Comment by zsol** [comment]
Fair enough. If anyone's interested in tackling this, here are a bunch of pointers:

There's best-effort recursion depth checking that's defined here:
https://github.com/Instagram/LibCST/blob/e9bad94d5868ba5a3919cc34e26665d40cdf9f91/native/libcst/src/parser/grammar.rs#L88

It's checked in a couple of places, but we could definitely add more checks.

More context in https://github.com/Instagram/LibCST/issues/619, another example in https://github.com/Instagram/LibCST/issues/980
